### DISCUSSION TITLE
Pruning of uBO filter*.txt lists

### DISCUSSION THREAD

**Discussion by ryanbr** [uBlockOrigin/uAssets#22698] [General]
Thoughts on limiting the uBO filters*.txt lists. (to avoid bloat). I suspect there is many sites where filters aren't needed any more or need updating. 

**Currently:**
- Currently, every year a new list is created with filters-[year].txt
- New rules are added to the new list, and updating on the other lists

**Proposed:**
- Every year a new list is created with filters-[year].txt (no change).
- At the start/end of the year (or whenever), Review the oldest list at the time (filters.txt, filters-2020.txt etc..), and go through each of the filters. (git blame/git log)
- Any adblock rules still confirmed current is moved into the latest list filters-[year].txt.
- The oldest filters list could be null'd eventually.

**Why:**
- There is rules that are many years old, which may or may not apply any more. 
- Will avoid uBO list bloat over time as much as possible, and check if the filters need to be removed, changed or what ever. 
- Keeps filter lists more reviewed more frequently, well at least yearly.

**Comment by ryanbr** [comment]
I know that this isn't the easiest of tasks.

But reviewing the oldest rules. and moving them into the updated list should be a good idea. Keeps the list more optimised, and more managable.

Something similar is done in Easylist, and its still manual (where's AI when you need it). Using `git blame` and `git log`. viewing the oldest additions and reviewing the commits. View the site, see if still applies and remove or change the filters.

**Reply by gorhill** [reply]
> At the start/end of the year (or whenever), Review the oldest list at the time (filters.txt, filters-2020.txt etc..), and go through each of the filters. (git blame/git log)

The idea I had in mind originally was to just drop the oldest list and just revise/move filters from high traffic sites in it -- I don't think it's worth spending time on 6-years old filters from low traffic sites. Also back then we didn't have report tool, this way we can be sure that should there be filters still valid in a dropped old file, we would hear about it soon enough.

**Reply by ryanbr** [reply]
Yeah high traffic vs low traffic sites, every user reporting a site thinks its important regardless of its ranking. Though it doesn't take into account sites that are new-ish (spun up new servers for pirate streams for example). I don't think that battle of only the top million, 100k sites would work, esp on the more legally questionable sites.

Just doing a quick check on filters.txt; Do something similar on easylist (on each sepecific easylist_* file), visit and see what filter applies;

```
git blame filters.txt | grep 2017- | less
git blame filters.txt | grep 2018- | less
git blame filters.txt | grep 2019- | less
...
```

Moving them into another list is optional, but sort of forces a review of the older rules. Up to devs, but any rules older than 2021 should be reviewed. How applicable is it, can it be removed or changed. Did the cosmetic was simply removed by the site, something a report tool won't pick up on

**Reply by Yuki2718** [reply]
I also don't like stacking up -20xx.txt as it's inconvenient e.g. when I check dead domain. But maybe filters.txt should be kept as a base and eventually should include only rules for high-traffic sites such as yt, long-lived rules, and maybe common rules for popMagic, Propeller, etc. But as you expected, it's not an easy task and would need year- or at least months-scale time to complete. I have been cleaning up filters.txt when I found time, but mosf of cleanup are done solely by mapx except for recent notable work by huyszn.

**Reply by Yuki2718** [reply]
@ryanbr As you talked about a plan to add _specific_uBO.txt to EL in EL chat, it's probably best to sync with. On the process of reviewing old rules we will find rules that can be moved to EL.

**Reply by ryanbr** [reply]
The idea I guess is to remove the amount of filters and the amount of files over time as they're reviewed.

Having less files will be easier for maintence. I think a good goal is to remove dead filters in filers.txt and migrate working rules to 2024 (with any references in the comment). Long lived rules can still be applied, no need to treat them differently tbh, and referenced in a comment in a filename.

```
! migrated from filters.txt. Checked 28/02/204 
! https://github.com/uBlockOrigin/uAssets/issues/xxxxx
somedomain.com##+js(fsfsfsdfs)
```

Even if we start small and just aim for removing `filters.txt` this year, and then next year `filters-2020.txt`, and so forth. Will mean that the filters-* won't get out of control and much easier to maintain.

**Reply by ryanbr** [reply]
> @ryanbr As you talked about a plan to add _specific_uBO.txt to EL in EL chat, it's probably best to sync with. On the process of reviewing old rules we will find rules that can be moved to EL.

Sure, but depends on how receptive uBO authors are. Easyprivacy specific uBO is there, but whether we import from uBO uAssets and uBO authors are receptive of removing rules covered in our lists is another thing.

**Reply by Yuki2718** [reply]
> with any references in the comment

Maybe add lines like
```
! ---------- Rules from filters.txt ----------

! ---------- Rules from filters-2020.txt ----------
```
to the bottom of 2024.txt and move to them will be fine. Some rules for recent issues can be moved to common section of 2024.txt though.

**Reply by ryanbr** [reply]
Sure maybe add a START and END

**Reply by Yuki2718** [reply]
https://github.com/uBlockOrigin/uAssets/commit/d5c3b06b0d775911dad18f1e9faa3ebc92ab6b05
Chose the top than the bottom so not to interfere with our practice of adding new rules. Next year we can move the whole contents to 2025.txt with new lines of `Rules from filters-2024.txt` added. Also created `Movable to EL` section. Please note EL addresses anti-adb only for adult sites, shortners/file hostings, and streaming/torrent/comic sites (see https://github.com/easylist/easylist). No need to move rules needing constant adjustment though, as members other than me don't have write access to EL.

**Reply by gorhill** [reply]
I think only filters from a set number of years should be moved to current year, like 2024 - 6 years ago = <=2018 , and once moved, they become "valid as of 2024", they will be checked again in 2030 since 2024 will become 6 years old. The idea is to garbage collect only old filters, not the one created in recent time.

**Reply by Yuki2718** [reply]
I don't see problem for moving rules like
```
! https://forums.lanik.us/viewtopic.php?f=62&t=42076
hotbabes.tv##+js(aopr, loadTool)
candid.tube,filesamba.*,hotbabes.tv,purelyceleb.com##+js(aopr, _wm)
cambabe.*,camgirlbay.net##+js(acs, onload)
cambabe.*##+js(acs, crakPopInParams)
cambabe.*##.topad
cambabe.*##.place
purelyceleb.com##.embed-container
##.exo-horizontal
```
in `filters.txt` to 2024.txt.

**Reply by gorhill** [reply]
Sure, all of `filters.txt` should be assumed to be very old. Just pointing out you created sections for 2020-2023, I don't think we should mind such "recent" filters,  and also:

> Next year we can move the whole contents to 2025.txt

The whole content of what? I am getting confused here. The way I see it to keep the workload as small and manageable is to just mind old filters (after we define "old"), and once an old filter has been validated as still being needed, we move it to current year list, then we do not validate the content of that list until it also become old in a few years.

**Reply by Yuki2718** [reply]
All right, got it. I was thinking to keep only `filters.txt` and the latest year's list. But maybe 6 is too long, I guess 3 or 5 years will be suffice.

**Comment by Yuki2718** [comment]
Initial commit https://github.com/uBlockOrigin/uAssets/commit/748b0464bc4136fc9d52dd72691c979e55236843
This actualy can be moved to EL once _specific_uBO.txt is added, but I haven't heard from other members about this.
